qcodes/instrument_drivers/stanford_research/SR830.py

```diff
@@ -1,15 +1,76 @@
 from functools import partial
 import numpy as np
-from typing import Any
+from typing import Any, Union
 
 from qcodes import VisaInstrument
-from qcodes.instrument.parameter import ArrayParameter, ParamRawDataType
-from qcodes.utils.validators import Numbers, Ints, Enum, Strings
+from qcodes.instrument.parameter import ArrayParameter, ParamRawDataType, ParameterWithSetpoints, Parameter
+from qcodes.utils.validators import Numbers, Ints, Enum, Strings, Arrays
 
 from typing import Tuple
 import time
 
 
+class ChannelTrace(ParameterWithSetpoints):
+    """
+    Parameter class for the two channel buffers
+    """
+
+    def __init__(self, name: str, channel: int, **kwargs: Any) -> None:
+        """
+        Args:
+            name: The name of the parameter
+            channel: The relevant channel (1 or 2). The name should
+                should match this.
+        """
+        super().__init__(name, **kwargs)
+
+        self._valid_channels = (1, 2)
+
+        if channel not in self._valid_channels:
+            raise ValueError('Invalid channel specifier. SR830 only has '
+                             'channels 1 and 2.')
+
+        if not isinstance(self.root_instrument, SR830):
+            raise ValueError('Invalid parent instrument. ChannelBuffer '
+                             'can only live on an SR830.')
+
+        self.channel = channel
+        self.update_unit()
+
+    def update_unit(self) -> None:
+        assert isinstance(self.root_instrument, SR830)
+        params = self.root_instrument.parameters
+        if params[f'ch{self.channel}_ratio'].get() != 'none':
+            self.unit = '%'
+        else:
+            disp = params[f'ch{self.channel}_display'].get()
+            if disp == 'Phase':
+                self.unit = 'deg'
+            else:
+                self.unit = 'V'
+                self.label = self.root_instrument.name + '_' + disp
+
+    def get_raw(self) -> ParamRawDataType:
+        """
+        Get command. Returns numpy array
+        """
+        assert isinstance(self.root_instrument, SR830)
+        N = self.root_instrument.buffer_npts()
+        if N == 0:
+            raise ValueError('No points stored in SR830 data buffer.'
+                             ' Can not poll anything.')
+
+        # poll raw binary data
+        self.root_instrument.write(f'TRCL ? {self.channel}, 0, {N}')
+        rawdata = self.root_instrument.visa_handle.read_raw()
+
+        # parse it
+        realdata = np.fromstring(rawdata, dtype='<i2')
```

Comment:
I don’t have any strong feeling about this; if you thing the query_binary_data method  I will try it out

Reply:
nah, if this works, then let's not bother. it works and the code is clear. my comment was more to mention the query method

Reply:
read_raw will only read up to a given chunk size so using `query_binary_values` is definitely less error prone. In this example it may not matter but as I understand it `read_raw` really never should be used in production code since it disables many of the safeties of the more high level functions